Atom launch scenario descriptions

smashwilson opened the pull request:
### Requirements for Contributing Documentation

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only contribute documentation (for example, markdown files or API docs). To contribute other changes, you must use a different template. You can see all templates at https://github.com/atom/atom/tree/master/.github/PULL_REQUEST_TEMPLATE.

### Description of the Change

Exhaustively document Atom launch scenarios and behavior across versions 1.35.1, 1.36.0, and speculative behavior for a hypothetical 1.36.1.

See the discussion in #19147 for additional feedback and discussion.

### TODO

- [x] Describe common scenarios when launching from the command line or via file manager
- [x] Split `[/a] []` sections to cover order-dependent launch behavior
- [ ] Describe the zero-argument scenarios (i.e., launch from CLI as `atom` or by clicking on dock/start menu)
- [ ] Describe how the `core.restorePreviousWindowsOnStart` setting affects these scenarios

### Release Notes

_N/A_


-----
[View rendered docs/launch-behavior.md](https://github.com/atom/atom/blob/aw/document-launch-behaviors/docs/launch-behavior.md)

smashwilson committed:
Shorthand key and CLI action tables

smashwilson committed:
Markdown tables are hard okay

smashwilson committed:
Chart out file manager operations

smashwilson committed:
Mark changes with :new:

jasonrudolph reviewed (commented):
:wave: Hey @smashwilson: Thanks for putting this together. I really think this is gonna help us have a more clear grasp of this behavior.

I ran through all of the scenarios in Atom 1.33.1. I wanted to try things out in recent, but not current, version of Atom so that we could understand how things have worked for most of Atom's lifetime. (I wanted to avoid 1.34.x and 1.35.x, because there were a few changes to file-opening behavior in that timeframe: https://github.com/atom/atom/pull/17529, https://github.com/atom/atom/issues/18941, https://github.com/atom/atom/issues/18835, https://github.com/atom/atom/pull/18855)

In a handful of scenarios, the behavior I observed differs from the behavior documented in this pull request. When you have a moment, could you look over my findings and let me know if you're seeing the behavior I'm seeing?

---

Note: I haven't yet reviewed the proposed behavior. I'll get started on that next. :-)

jasonrudolph commented:
@smashwilson: Just FYI: I added a couple tasks to the pull request body so that we don't lose track of a couple other variations you mentioned in chat (zero-argument scenarios and the impact of the `core.restorePreviousWindowsOnStart` setting).

I think we can have a really good discussion based on the content that we already have in this pull request, but before we wrap this up, I think the document should cover these other scenarios to avoid potential surprises.

jasonrudolph committed:
Apply suggestions from code review

Looks like I forgot that old Atom versions would "collapse" new open actions into a most-recently-focused empty window, which changed a few cases.

Also it looks like we _did_ always imply `--add` when given a file path... ? That's unexpected.

Co-Authored-By: smashwilson <smashwilson@github.com>

smashwilson committed:
Order-dependent rows in the existing tables

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+| --- | --- | --- | --- |
+| `atom /a/1.md`              | [/a 1.md] []      | [/a 1.md] []          | [/a 1.md] []       |
+| `atom /a`                   | [/a] []           | [/a] []               | [/a] []            |
+| `atom /b/3.md`              | [/a] [] [/b 3.md] | [/a] [] [ 3.md] :new: | [/a] [ 3.md] :new: |
```
👍 

Resolved in b9a1d91059cc06ccc9470301d4db48e8f11c3337.

smashwilson replied:
`[] [/a]` cases are now below :point_down:

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+| `atom /a/1.md`              | [/a 1.md] []      | [/a 1.md] []          | [/a 1.md] []       |
+| `atom /a`                   | [/a] []           | [/a] []               | [/a] []            |
+| `atom /b/3.md`              | [/a] [] [/b 3.md] | [/a] [] [ 3.md] :new: | [/a] [ 3.md] :new: |
+| `atom /b`                   | [/a] [] [/b]      | [/a] [] [/b]          | [/a] [] [/b]       |
```
Resolved in b9a1d91059cc06ccc9470301d4db48e8f11c3337.

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+| --- | --- | --- | --- |
+| Click on `/a/1.md` | [/a 1.md] []      | [/a 1.md] []          | [/a 1.md] []       |
+| Click on `/a`      | [/a] []           | [/a] []               | [/a] []            |
+| Click on `/b/3.md` | [/a] [] [/b 3.md] | [/a] [] [ 3.md] :new: | [/a] [ 3.md] :new: |
```
Resolved in b9a1d91059cc06ccc9470301d4db48e8f11c3337.

smashwilson replied:
The `[] [/a]` cases are now their own table :ok_hand:

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+
+| Action | Outcome (<=1.35.1) | Outcome (1.36.0) | Outcome (1.36.1) |
+| --- | --- | --- | --- |
+| Click on `/a/1.md` | [/a 1.md] | [ 1.md] :new: | [ 1.md] |
```
I think this change is probably OK. However, it is a _change_, and this change wasn't an explicit goal of #18608, so I think we should be open to restoring the previous behavior if there's a compelling argument for doing so.

smashwilson replied:
Yeah, I'm on the fence about this one. It feels like we'd want to avoid indexing the directory here too, so that we launch quickly when clicking on a file in your home directory just as we do when you run `atom ~/file.txt` from the CLI. But I could see an argument that user expectations may differ?

jasonrudolph replied:
> But I could see an argument that user expectations may differ?

I don't personally anticipate user expectations differing here. I personally prefer this change, but I'll be curious to see if it gets any pushback.

smashwilson committed:
Add charts for flipped window orders

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+| Action | Outcome (<=1.35.1) | Outcome (1.36.0) | Outcome (1.36.1) |
+| --- | --- | --- | --- |
+| Click on `/a/1.md` | [/a 1.md] | [] [ 1.md] :new: | [ 1.md] :new: |
+| Click on `/a`      | [/a]      | [] [/a]          | [] [/a]       |
```
I think 1.36.1 should restore the previous behavior. This change in behavior wasn't an explicit goal of #18608 (AFAIK), and since launching a new window is slow, this change arguably degrades the user experience.

```suggestion
| Click on `/a`      | [/a]      | [] [/a]          | [/a]       |
```

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+| --- | --- | --- | --- |
+| Click on `/a/1.md` | [/a 1.md]      | [/a 1.md]          | [/a 1.md]    |
+| Click on `/a`      | [/a]           | [/a]               | [/a]         |
+| Click on `/b/3.md` | [/a 3.md]      | [/a] [ 3.md] :new: | [/a] [ 3.md] |
```
I think 1.36.1 should restore the previous behavior. This change in behavior wasn't an explicit goal of #18608 (AFAIK), and since launching a new window is slow, this change arguably degrades the user experience.

```suggestion
| Click on `/b/3.md` | [/a 3.md]      | [/a] [ 3.md] :new: | [/a 3.md] |
```

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,137 @@
+
+| Action | Outcome (<=1.35.1) | Outcome (1.36.0) | Outcome (1.36.1) |
+| --- | --- | --- | --- |
+| Click on `/a/1.md` | [/a 1.md] | [] [ 1.md] :new: | [ 1.md] :new: |
```
I think this change is probably OK. However, it is a _change_, and this change wasn't an explicit goal of #18608, so I think we should be open to restoring the previous behavior if there's a compelling argument for doing so.

jasonrudolph reviewed (changes requested):
@smashwilson: I noted a couple suggested changes to the file manager integration actions. Would you mind looking them over and letting me know what you think?

I'll look through the CLI actions shortly.

smashwilson committed:
Verify 1.36.0 tables and update emoji

jasonrudolph committed:
Apply suggestions from code review

Co-Authored-By: smashwilson <smashwilson@github.com>

smashwilson committed:
Verify, correct, and re-emoji the [/a] CLI section

jasonrudolph reviewed (changes requested):
@smashwilson: I noted a few suggested changes to the CLI actions. Would you mind looking them over and letting me know what you think?

smashwilson committed:
Verify, correct, re-emoji [/a] [] CLI section

smashwilson committed:
Populate the [] [/a] CLI section

smashwilson committed:
Verify and populate the 1.36.0 file manager outcomes

jasonrudolph committed:
Apply suggestions from code review

Co-Authored-By: smashwilson <smashwilson@github.com>

smashwilson committed:
Re-emoji the 1.36.1 columns

smashwilson committed:
Re-verify and fix some behavior on a suspicious entry

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,165 @@
+| `atom /a/1.md`              | [] [/a 1.md]      | [] [/a 1.md]              | [] [/a 1.md]          |
+| `atom /a`                   | [] [/a]           | [] [/a]                   | [] [/a]               |
+| `atom /b/3.md`              | [] [/a 3.md]      | [] [/a] [ 3.md] :warning: | [] [/a 3.md] :bug:    |
+| `atom /b`                   | [] [/a] [/b]      | [] [/a] [/b]              | [/b] [/a] :bug: :new: |
```
This change feels like a "nice-to-have" to me. If we can make it work easily, then I think it's a nice improvement. Otherwise, I recommend that we defer it for a future pull request, since this change wasn't an explicit goal of #18608.

smashwilson replied:
👍 Makes sense to me. I _think_ it'll actually be easier to implement this than preserving the old inconsistent behavior, but I won't know for sure until I'm all up in AtomApplication again.

jasonrudolph commented on docs/launch-behavior.md:
```diff
@@ -0,0 +1,165 @@
+| Click on `/a/1.md` | [] [/a 1.md] | [] [/a 1.md]              | [] [/a 1.md]             |
+| Click on `/a`      | [] [/a]      | [] [/a]                   | [] [/a]                  |
+| Click on `/b/3.md` | [] [/a 3.md] | [] [/a] [ 3.md] :warning: | [ 3.md] [/a] :bug: :new: |
+| Click on `/b`      | [] [/a] [/b] | [] [/a] [/b]              | [/b] [/a] :bug: :new:    |
```
This change feels like a "nice-to-have" to me. If we can make it work easily, then I think it's a nice improvement. Otherwise, I recommend that we defer it for a future pull request, since this change wasn't an explicit goal of #18608.

jasonrudolph reviewed (commented):
This looks great, @smashwilson! :zap::bow: I left a couple comments inline below, but overall, I'm 👍 on everything here. Thanks for writing this up! 

Do you have any thoughts on how this document could address the "zero-argument scenarios" and 
the effects of the `core.restorePreviousWindowsOnStart` setting?

smashwilson committed:
Consistency between file manager and CLI

We had one inconsistency in behavior between the file manager and CLI opening scenarios - the one where you open a file outside of any project roots when an empty window is present, but not the last-focused window. Let's bring these in line to make things more predictable.

smashwilson commented:
Closing this to tidy up pull request clutter. The source of truth for Atom's launch behavior is now the descriptions in the main process test suite:

https://github.com/atom/atom/blob/master/spec/main-process/atom-application.test.js
